# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1711900088 -7200
# Parent  6647479dd6fc6287d90bf76fa1d85960ea506590
Feature: Shift+"Mark Folder Read" also marks sub-folders read.

Requested in https://bugzilla.mozilla.org/show_bug.cgi?id=500762 in 2009.

diff --git a/mail/base/content/about3Pane.js b/mail/base/content/about3Pane.js
--- a/mail/base/content/about3Pane.js
+++ b/mail/base/content/about3Pane.js
@@ -660,8 +660,34 @@ var folderPaneContextMenu = {
         if (folder.flags & Ci.nsMsgFolderFlags.Virtual) {
           topChromeWindow.MsgMarkAllRead(
             VirtualFolderHelper.wrapVirtualFolder(folder).searchFolders
           );
+        } else if (event.shiftKey) {
+          // Mark folder and subfolders read.
+          function recurse(f) {
+            let subFolders;
+            try {
+              subFolders = f.subFolders;
+            } catch (ex) {
+              console.error(
+                new Error(`Unable to access the subfolders of ${f.URI}`, {
+                  cause: ex,
+                })
+              );
+            }
+            if (!subFolders?.length) {
+              return;
+            }
+
+            for (const sf of subFolders) {
+              folders.push(sf);
+              recurse(sf);
+            }
+          }
+
+          const folders = [folder];
+          recurse(folder);
+          topChromeWindow.MsgMarkAllRead(folders);
         } else {
           topChromeWindow.MsgMarkAllRead([folder]);
         }
         break;
